go/store/prolly/node.go

```diff
@@ -0,0 +1,218 @@
+//   | Offset 1 | Offset 2 | ... | Offset N |
+//   +----------+----------+-----+----------+
+//
+//   Cumulative Count:
+//      The cumulative count is the total number of items in the subtree rooted at
+//      this node. For leaf nodes, cumulative count is the same as node count.
+//   +---------------------------+
+//   | Cumulative Count (uint48) |
+//   +---------------------------+
+//
+//   Node Count:
+//      Node count is the number of items in this node.
+//   +---------------------+
+//   | Node Count (uint16) |
+//   +---------------------+
+//
+//   Tree Level:
+//      Tree Level is the height of this node within the tree. Leaf nodes are
+//      level 0, the first level of internal nodes is level 1.
+//   +--------------------+
+//   | Tree Level (uint8) |
+//   +--------------------+
+//
+//   Note: the current Node implementation is oriented toward implementing Map
+//   semantics. However, Node could easily be modified to support Set semantics,
+//   or other collections.
+//
+type Node []byte
+
+type nodeItem []byte
+
+func (i nodeItem) size() val.ByteSize {
+	return val.ByteSize(len(i))
+}
+
+type nodePair [2]nodeItem
+
+func (p nodePair) key() nodeItem {
+	return p[0]
+}
+
+func (p nodePair) value() nodeItem {
+	return p[1]
+}
+
+func makeProllyNode(pool pool.BuffPool, level uint64, items ...nodeItem) (node Node) {
+	var sz uint64
+	for _, item := range items {
+		sz += uint64(item.size())
+
+	}
+	count := len(items)
+
+	if sz > maxNodeDataSize {
+		panic("items exceeded max chunk data size")
+	}
+
+	pos := val.ByteSize(sz)
+	pos += val.OffsetsSize(count)
+	pos += cumulativeCountSize
+	pos += nodeCountSize
+	pos += treeLevelSize
+
+	node = pool.Get(uint64(pos))
+
+	cc := countCumulativeItems(level, items)
+	writeCumulativeCount(node, cc)
+	writeItemCount(node, count)
+	writeTreeLevel(node, level)
+
+	pos = 0
+	offs, _ := node.offsets()
+	for i, item := range items {
+		copy(node[pos:pos+item.size()], item)
+		offs.Put(i, pos)
+		pos += item.size()
+	}
+
+	return node
+}
+
+func countCumulativeItems(level uint64, items []nodeItem) (c uint64) {
+	if level == 0 {
+		return uint64(len(items))
+	}
+
+	for i := 1; i < len(items); i += 2 {
+		c += metaValue(items[i]).GetCumulativeCount()
+	}
+	return c
+}
+
+func (nd Node) getItem(i int) nodeItem {
+	offs, itemStop := nd.offsets()
+	start, stop := offs.GetBounds(i, itemStop)
+	return nodeItem(nd[start:stop])
+}
+
+func (nd Node) getPair(i int) (p nodePair) {
+	offs, itemStop := nd.offsets()
+	start, stop := offs.GetBounds(i, itemStop)
+	p[0] = nodeItem(nd[start:stop])
+	start, stop = offs.GetBounds(i+1, itemStop)
+	p[1] = nodeItem(nd[start:stop])
+	return
+}
+
+func (nd Node) size() val.ByteSize {
+	return val.ByteSize(len(nd))
+}
+
+func (nd Node) level() int {
+	sl := nd[nd.size()-treeLevelSize:]
+	return int(val.ReadUint8(sl))
+}
+
+func (nd Node) nodeCount() int {
+	stop := nd.size() - treeLevelSize
+	start := stop - nodeCountSize
+	return int(val.ReadUint16(nd[start:stop]))
```

Comment:
the need for (de)serialization here and below motivated me to make the codec functions public. I considered even splitting them out to their own package.